pkg/ddl/sanity_check.go

```diff
@@ -113,24 +113,22 @@ func expectedDeleteRangeCnt(ctx delRangeCntCtx, job *model.Job) (int, error) {
 		}
 		return len(args.OldPhysicalTblIDs), nil
 	case model.ActionAddIndex, model.ActionAddPrimaryKey:
-		indexID := make([]int64, 1)
-		ifExists := make([]bool, 1)
-		isGlobal := make([]bool, 0, 1)
-		var partitionIDs []int64
-		if err := job.DecodeArgs(&indexID[0], &ifExists[0], &partitionIDs); err != nil {
-			if err := job.DecodeArgs(&indexID, &ifExists, &partitionIDs, &isGlobal); err != nil {
-				var unique bool
-				if err := job.DecodeArgs(&unique); err == nil {
-					// The first argument is bool means nothing need to be added to delete-range table.
-					return 0, nil
-				}
-				return 0, errors.Trace(err)
+		// In the previous code, we would first try to decode args using finished arguments.
+		// If decode failed, then we assume that the job is not finished and decode it again.
+		// After switching to v2, we use IsFinishedArg flag to distinguish these two cases.
```

Comment:
we can determine which function to call by check whether job state is `synced`

Reply:
`IsFinishedArg` is added in this PR. You mean we can call `GetAddIndexArgs` directly if `job.State != JobStateSynced`?

Reply:
ok for now. but the comment `After switching to v2, we use IsFinishedArg flag to distinguish these two cases.` seems irrelevant

Reply:
Because in v2, `GetFinishedAddIndexArgs` will return error if `IsFinishedArg` is not set.
https://github.com/pingcap/tidb/blob/b6bfc48e35db9fe2d792ccadf62d610e4d8bf3d9/pkg/meta/model/job_args.go#L1634-L1636

Reply:
we shouldn't depends on it for history job,  this flag is not persisted for V1

Reply:
If we assume that history job always used v1, `GetFinishedAddIndexArgs` will fail to decode and return an error if finished arguments is not set.